eZ/Publish/Core/Persistence/Legacy/SharedGateway/GatewayFactory.php

```diff
@@ -0,0 +1,51 @@
+<?php
+
+/**
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace eZ\Publish\Core\Persistence\Legacy\SharedGateway;
+
+use Doctrine\DBAL\Connection;
+use Traversable;
+
+/**
+ * Builds Shared Gateway object based on the database connection.
+ *
+ * @internal For internal use by Legacy Storage Gateways.
+ */
+final class GatewayFactory
+{
+    /** @var \eZ\Publish\Core\Persistence\Legacy\SharedGateway\Gateway */
+    private $fallbackGateway;
+
+    /** @var \Traversable|\eZ\Publish\Core\Persistence\Legacy\SharedGateway\Gateway[] */
+    private $gateways;
+
+    public function __construct(Gateway $fallbackGateway, Traversable $gateways)
```

Comment:
> This will simplify `GatewayFactory` construction in
> 
> * https://github.com/ezsystems/ezpublish-kernel/pull/2845/files#diff-577e34af8f9df09e67805be30558918cR135-R142
> * https://github.com/ezsystems/ezpublish-kernel/pull/2845/files#diff-a221d9ac1d4eda8d60063ddd01799e13R37-R48

Refactored and squashed into 75ae1ce.